libkbfs/tlf_handle_resolve.go

```diff
@@ -0,0 +1,560 @@
+// the error. Usually, you want to treat this as a symlink to the name
+// to try.
+//
+// NoSuchNameError: Returned when public is set and the given folder
+// has no public folder.
+func ParseTlfHandle(
+	ctx context.Context, kbpki KBPKI, name string, public bool) (
+	*TlfHandle, error) {
+	// Before parsing the tlf handle (which results in identify
+	// calls that cause tracker popups), first see if there's any
+	// quick normalization of usernames we can do.  For example,
+	// this avoids an identify in the case of "HEAD" which might
+	// just be a shell trying to look for a git repo rather than a
+	// real user lookup for "head" (KBFS-531).  Note that the name
+	// might still contain assertions, which will result in
+	// another alias in a subsequent lookup.
+	// This also contains an offline check for canonicality and
+	// that a public folder has readers.
+	writerNames, readerNames, extensionSuffix, err :=
+		splitAndNormalizeTLFName(name, public)
+	if err != nil {
+		return nil, err
+	}
+
+	changesCh := make(chan struct{}, 1)
+	writers := make([]resolvableUser, len(writerNames))
+	for i, w := range writerNames {
+		writers[i] = resolvableAssertionWithChangeReport{
+			resolvableAssertion{kbpki, w, keybase1.UID("")}, changesCh}
+	}
+	readers := make([]resolvableUser, len(readerNames))
+	for i, r := range readerNames {
+		readers[i] = resolvableAssertionWithChangeReport{
+			resolvableAssertion{kbpki, r, keybase1.UID("")}, changesCh}
+	}
+
+	var extensions []TlfHandleExtension
+	if len(extensionSuffix) != 0 {
+		extensions, err = ParseTlfHandleExtensionSuffix(extensionSuffix)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	h, err := makeTlfHandleHelper(ctx, public, writers, readers, extensions)
+	if err != nil {
+		return nil, err
+	}
+
+	if !public {
+		currentUsername, currentUID, err := kbpki.GetCurrentUserInfo(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if !h.IsReader(currentUID) {
+			return nil, ReadAccessError{currentUsername, h.GetCanonicalName(), public}
+		}
+	}
+
+	if extensionSuffix != "" {
+		extensionList := tlfHandleExtensionList(extensions)
+		sort.Sort(extensionList)
+		var canonExtensionString = extensionList.Suffix()
+		if canonExtensionString != TlfHandleExtensionSep+extensionSuffix {
+			return nil, TlfNameNotCanonical{name, string(h.GetCanonicalName())}
+		}
+	}
+
+	select {
+	case <-changesCh:
+	default:
+		// No changes were performed because of resolver.
+		return h, nil
+	}
+
+	// Otherwise, identify before returning the canonical name.
+	err = identifyHandle(ctx, kbpki, kbpki, h)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, TlfNameNotCanonical{name, string(h.GetCanonicalName())}
+}
+
+// ParseTlfHandlePreferred returns TlfNameNotCanonical if not
+// in the preferred format.
+func ParseTlfHandlePreferred(
+	ctx context.Context, kbpki KBPKI, name string, public bool) (
+	*TlfHandle, error) {
+	h, err := ParseTlfHandle(ctx, kbpki, name, public)
+	if err != nil {
+		return nil, err
```

Comment:
If this returns `TlfNameNotCanonical`, the `NameToTry` will not be the preferred name, so it will end up making a symlink to another symlink.  That's not a big deal, but it might be nicer to overwrite the `NameToTry` with the preferred name in that case if we can.